Second PT1 on DTerm

ctzsnooze opened the pull request:
This PR allows the user to enable a second, independently configurable set point, PT1 type first order low-pass filter on DTerm.

This is very useful as a means of controlling because most noise in most logs arises from D, not P.

User sets dterm_lowpass_2 in CLI to the desired cut point.  Double or triple the lower, standard Dterm lowpass is a good starting point.  Default is 200.

The default is set to on, at twice the normal Dterm setpoint, and the default Dterm filter type is changed from biquad to PT1.  

This provides greater Dterm cut than a single PT1, with twice the steepness of cut above the second setpoint.  Modelling and logging shows significant reductions in higher frequency Dterm noise with only minor additional delay.  Having a configurable setpoint allows the user to tune the least filtering that provides adequate noise reduction.

The biquad will provide very effective filtering but greater delay.

The following images show logs of four flights, the first panel shows that P noise is about the same on each, the second shows D noise with, from left to right, a single D PT1 at 100, and then the effect of adding a second D PT1 in series at 300, 400 and 500Hz.  The second PT1 provides configurable and significant reductions in D noise from around to above its setpoint.

![amount of p in each trace](https://user-images.githubusercontent.com/11737748/37186564-4fe31ca4-239a-11e8-97dd-78427deba5c8.jpg)

![amount of d in each trace](https://user-images.githubusercontent.com/11737748/37186644-a2093fc2-239a-11e8-83cd-f34d0dd3150d.jpg)

A spreadsheet that allows visualisation of the effect on random noise and other inputs is here https://www.dropbox.com/s/36c8ec55c3hmqoa/DualDFilters1.xlsx?dl=0

DieHertz commented on src/main/flight/pid.c:
```diff
@@ -171,6 +172,8 @@ static FAST_RAM filterApplyFnPtr dtermNotchFilterApplyFn;
 static FAST_RAM filterApplyFnPtr dtermLpfApplyFn;
 static FAST_RAM void *dtermFilterLpf[2];
+static FAST_RAM filterApplyFnPtr dtermLpf2ApplyFn;
+static FAST_RAM void *dtermLpf2Filter[2];
```
Please make the naming uniform with existing code, e.gm `dtermFilterLpf2` instead.
Or better, let's rename old and new to `dtermLowpass` and `dtermLowpass2`. Why have the world filter twice in one name? :)

fujin replied:
`lowpass`/`lowpass2` is also the naming I went with in the dual-stage PR.

fujin replied:
makes me wonder if we should be using `FAST_RAM` in gyroSensor for similar storage as well...

DieHertz replied:
Don't dive into that, I'm working on it as a part of F7 optimization branch

ctzsnooze replied:
OK will change to to dtermLowpass and dtermLowpass2

DieHertz commented on src/main/interface/settings.c:
```diff
@@ -637,6 +637,7 @@ const clivalue_t valueTable[] = {
 
 // PG_PID_PROFILE
     { "dterm_lowpass_type",         VAR_UINT8  | PROFILE_VALUE | MODE_LOOKUP, .config.lookup = { TABLE_LOWPASS_TYPE }, PG_PID_PROFILE, offsetof(pidProfile_t, dterm_filter_type) },
+	{ "dterm_lowpass_2",              VAR_INT16  | PROFILE_VALUE, .config.minmax = { 0, 16000 }, PG_PID_PROFILE, offsetof(pidProfile_t, dterm_lpf_2_hz) },
```
Should be `dterm_lowpass2_type`

ctzsnooze replied:
With this implementation, the type can't be changed, it's set to a PT1, so the user enters the value in Hz.  

Maybe it should be 'dterm_lowpass_2_hz'

DieHertz replied:
No underscore before 2, please, to conform with gyro lowpass and notch

ctzsnooze replied:
done

fujin commented on src/main/flight/pid.c:
```diff
@@ -216,6 +219,18 @@ void pidInitFilters(const pidProfile_t *pidProfile)
     }
+    
+    //extra Dterm PT1 filter
+	static pt1Filter_t ExtraPt1FilterD;
```
remove 'extra',

will we ever allow others than pt1?

ctzsnooze replied:
I struggled with the name for this.  At the moment it's an additional Dterm filter, it is always a PT1, and the user can only enable/set the cutoff point.  Hence 'extra' (additional, but fewer letters), Dterm, PT1, Filter.

Happy for any particular name.

fujin replied:
It may be possible to refactor the new gyroInitLowpassLpf() function written to support dual/multi-stage dterm/yaw lowpass of varying type, we can do further renaming there after this is tested/merged.

for now, I suggest, e.g.:
```
static filter_t *dtermLowpassState
static pt1Filter_t dtermLowpass2State
```

`Extra` is fine for now as well if we want to revisit.

ctzsnooze replied:
Hmm it has to be different from the other names for the same thing

fujin commented on src/main/flight/pid.h:
```diff
@@ -75,6 +75,7 @@ typedef struct pidProfile_s {
 
     uint16_t yaw_lpf_hz;                    // Additional yaw filter when yaw axis too noisy
     uint16_t dterm_lpf_hz;                  // Delta Filter in hz
+    uint16_t dterm_lpf_2_hz;                // Extra PT1 Filter on D in hz
```
I'd been standardizing names and suggest the following:

`dterm_lowpass_hz`
`dterm_lowpass2_hz`

ctzsnooze replied:
yes no problem

ctzsnooze replied:
Actually no, decided to keep 'lpf' for the units themselves (the actual values), in CLI they are all 'lpf', so I will keep 'dterm_lpf_hz' etc here.  I think perhaps it's easier to read that way, 'lpf' is a value, 'lowpass' is a filter.

DieHertz replied:
I suggested renaming it in CLI as well to make it uniform with `gyro_lowpass_hz`

fujin commented on src/main/flight/pid.c:
```diff
@@ -511,6 +526,7 @@ void pidController(const pidProfile_t *pidProfile, const rollAndPitchTrims_t *an
             // apply filters
             float gyroRateFiltered = dtermNotchFilterApplyFn(dtermFilterNotch[axis], gyroRate);
             gyroRateFiltered = dtermLpfApplyFn(dtermFilterLpf[axis], gyroRateFiltered);
+			gyroRateFiltered = dtermLpf2ApplyFn(dtermLpf2Filter[axis], gyroRateFiltered);
```
whitespace/indentation here

ctzsnooze replied:
fixed (I think, pls check)

fujin commented on src/main/flight/pid.c:
```diff
@@ -216,6 +219,18 @@ void pidInitFilters(const pidProfile_t *pidProfile)
+    //extra Dterm PT1 filter
+	static pt1Filter_t ExtraPt1FilterD;
+	if (pidProfile->dterm_lpf_2_hz == 0 || pidProfile->dterm_lpf_2_hz > pidFrequencyNyquist) {
+		dtermLpf2ApplyFn = nullFilterApply;
```
better to set this regardless and then do the following block in the if(){} instead of if/else.

fujin commented on src/main/flight/pid.c:
```diff
@@ -105,9 +105,10 @@ void resetPidProfile(pidProfile_t *pidProfile)
         .pidSumLimitYaw = PIDSUM_LIMIT_YAW,
         .yaw_lpf_hz = 0,
         .dterm_lpf_hz = 100,    // filtering ON by default
+        .dterm_lpf_2_hz = 200,    // second PT1 ON by default
```
let's default to `0` (off) for now.

ctzsnooze replied:
OK

fujin replied:
We are not at release candidate stage presently but I would support/entertain a PR setting more aggressive defaults leading up to (or during) that phase. The general guidance I have received is to implement the functionality (default to off).

ctzsnooze replied:
no problem

fujin commented on src/main/flight/pid.c:
```diff
@@ -105,9 +105,10 @@ void resetPidProfile(pidProfile_t *pidProfile)
         .dterm_notch_hz = 260,
         .dterm_notch_cutoff = 160,
-        .dterm_filter_type = FILTER_BIQUAD,
+        .dterm_filter_type = FILTER_PT1,
```
changing defaults is generally not recommended at the same time as implementation

DieHertz replied:
I second that, let's collect more report first that for most people PT1+PT1 are better than Biquad

ctzsnooze replied:
OK

ledvinap commented:
@ctzsnooze : One property of chaining PT1 is that resulting filter Q is always 1/2 - critically damped. With higher Q (1/sqrt(2) for 2nd order Butterworth) it is possible to extend passband closer to cutoff frequency ...

fujin commented:
Looks really promising! I'm excited to test this both in combination with and instead of dual-stage gyro LPF's. Great work.

ctzsnooze commented:
@Ledivinap yes that's a good thing about second order filters.  I plan to compare a single biquad to dual separated PT1's, hopefully to establish cutoffs that actually give filtering equivalence in logs, and then to see if anyone can tell the difference in flight.  Theoretically will be less delay from separated PT1's, but also less filtering. At least this build allows a bit more fine tuning of how we filter D.

DieHertz commented on src/main/flight/pid.c:
```diff
@@ -105,6 +105,7 @@ void resetPidProfile(pidProfile_t *pidProfile)
         .pidSumLimitYaw = PIDSUM_LIMIT_YAW,
         .yaw_lpf_hz = 0,
         .dterm_lpf_hz = 100,    // filtering ON by default
+        .dterm_lpf_2_hz = 0,    // second Dterm LPF OFF by default
```
There's still an underscore before 2

DieHertz commented:
Thanks, there remains the point about underscore, and I think making lowpass naming conventions for gyro and Dterm uniform is beneficial.

ctzsnooze commented:
Thanks for all your feedback Fujin and DieHertz!  Definitely more readable.
I think I've implemented nearly all your suggestions, uploaded now, please check and see if anything else would help.
@DieHertz: For now I've left the 'value' type variable names in the form '_lpf_hz' to match current CLI nomenclature.  Changing them all would be a big change, people updating would get non-matching fields from CLI dumps.  Perhaps that could be separate pull request.
@Fujin: there is still a bit of whitespace problem in some places, I'll fix that on the next push.

DieHertz commented on src/main/interface/settings.c:
```diff
@@ -637,8 +637,8 @@ const clivalue_t valueTable[] = {
-	{ "dterm_lowpass_2",              VAR_INT16  | PROFILE_VALUE, .config.minmax = { 0, 16000 }, PG_PID_PROFILE, offsetof(pidProfile_t, dterm_lpf_2_hz) },
-    { "dterm_lowpass",              VAR_INT16  | PROFILE_VALUE, .config.minmax = { 0, 16000 }, PG_PID_PROFILE, offsetof(pidProfile_t, dterm_lpf_hz) },
+    { "dterm_lowpass_hz",              VAR_INT16  | PROFILE_VALUE, .config.minmax = { 0, 16000 }, PG_PID_PROFILE, offsetof(pidProfile_t, dterm_lpf_hz) },
+	{ "dterm_lowpass2_hz",              VAR_INT16  | PROFILE_VALUE, .config.minmax = { 0, 16000 }, PG_PID_PROFILE, offsetof(pidProfile_t, dterm_lpf2_hz) },
```
Looks like there's a redundant indent :)

ctzsnooze commented:
@DieHertz - Ah OK got it, removed those underscores.  Added _hz in CLI name for setpoint i.e. now dterm_lowpass_hz and dterm_lowpass2_hz

ctzsnooze commented:
Replaced 'lpf' within pid.c and related CLI names with 'lowpass', added _hz to CLI names for settings in hz, all should be consistent now, hope to have fixed whitespace too.

DieHertz commented on src/main/flight/pid.c:
```diff
@@ -207,55 +210,67 @@ void pidInitFilters(const pidProfile_t *pidProfile)
+	} else {
+		dtermLowpass2ApplyFn = (filterApplyFnPtr)pt1FilterApply;
+		for (int axis = FD_ROLL; axis <= FD_PITCH; axis++) {
+			dtermLowpass2[axis] = &dtermLowpass2State;
```
This falls short of using the same state for all axes, just like @fujin's PR before

DieHertz commented on src/main/flight/pid.c:
```diff
@@ -207,55 +210,67 @@ void pidInitFilters(const pidProfile_t *pidProfile)
+		}
+	}
+
+    static dtermLowpass_t dtermLowpassUnion;
```
Same here, this effectively makes state shared for all axes

ctzsnooze replied:
OK that was a mistake, unintended, need to fix that, sorry

ctzsnooze replied:
I checked, I've just used exactly the same method as for the yaw code and the other filter code as in release 3.3.   Hope there's nothing fundamentally wrong, but if there is, we will need to fix it.

DieHertz replied:
Yaw is just one axis, so it's not shared by definition.
In your code roll&pitch get shared Dterm state for lowpass.

DieHertz replied:
Lol at how it's done in Fujin's PR for gyro lowpass, you need to create state for PID filters in advance. I can help when I get home

ctzsnooze replied:
OK thanks I'll need your help with that.  I can see something he did in the gyro lowpass code, calculating pt1FilterGain, but looks like I'd have to implement a lot of his PR just to make this work!  Hopefully it's not going to be *that* difficult!  In the meantime, it seems to fly OK - out of curiosity, what would be the effect of this error?

DieHertz replied:
Coupling of roll and pitch, spikes and dips in roll will have effect on pitch at the same sample, and the same in pitch will have an effect on roll at the next sample

ctzsnooze replied:
Maybe the simplest, but ugliest, solution would be to run the code in pid.c on a per axis basis.  That would work.  Later after @fujin's PR he could use his gyro code to tidy up this PR?

ledvinap replied:
@DieHertz: https://github.com/betaflight/betaflight/pull/5427/files#diff-c5f9576172e273033324ba218524cc97R181

dtermLowpass_t contains arrays ...

DieHertz replied:
Okay then. I don't like this approach though.

ctzsnooze replied:
So it is OK?  I don't need to do anything now about that static issue?

DieHertz replied:
Yes, apologies

ledvinap replied:
3.3 is ok, your new code is not

ctzsnooze replied:
oh

ctzsnooze replied:
is it obvious what I need to do to fix it?

DieHertz replied:
Yes, add another static variable for lowpass2 state and use it for initialization of lowpass2

ledvinap replied:
@ctzsnooze : the problem is here: https://github.com/betaflight/betaflight/pull/5427/files#r173423911

ctzsnooze replied:
OK sorry I have to wait for actual code suggestion, too dumb to know how to do that.

ctzsnooze replied:
Would this work:

    //2nd Dterm Lowpass Filter      
    if (pidProfile->dterm_lowpass2_hz == 0 || pidProfile->dterm_lowpass2_hz > pidFrequencyNyquist) {
    	dtermLowpass2ApplyFn = nullFilterApply;
    } else {
    	static pt1Filter_t dtermLowpass2State[2];  
        dtermLowpass2ApplyFn = (filterApplyFnPtr)pt1FilterApply;
        for (int axis = FD_ROLL; axis <= FD_PITCH; axis++) {
            dtermLowpass2[axis] = &dtermLowpass2State[axis];
            pt1FilterInit(dtermLowpass2[axis], pidProfile->dterm_lowpass2_hz, dT);
        }        
    }

ledvinap replied:
dtermLowpass_t contains 2-element arrays for filter state (roll+pitch)
pt1Filter_t is single state

Use `static pt1Filter_t dtermLowpass2State[2];` or define type

ctzsnooze replied:
I just pushed this, hope it is now OK?

    //2nd Dterm Lowpass Filter   
    static pt1Filter_t dtermFilterLowpass2[2];     
    if (pidProfile->dterm_lowpass2_hz == 0 || pidProfile->dterm_lowpass2_hz > pidFrequencyNyquist) {
    	dtermLowpass2ApplyFn = nullFilterApply;
    } else {
        dtermLowpass2ApplyFn = (filterApplyFnPtr)pt1FilterApply;
        for (int axis = FD_ROLL; axis <= FD_PITCH; axis++) {
            dtermLowpass2[axis] = &dtermFilterLowpass2[axis];
            pt1FilterInit(dtermLowpass2[axis], pidProfile->dterm_lowpass2_hz, dT);
        }        
    }

DieHertz commented:
I can help you once I get home ;-)

ctzsnooze commented:
Needed to update name references in various other places also, e.g. pid.h, CMS, blackbox, etc.
Haven't touched abbreviations on-screen in OSD, nor have changed the abbreviations for lpf as in hardware gyro_lpf thats for someone else  :-)

ledvinap commented on src/main/flight/pid.h:
```diff
@@ -73,8 +73,9 @@ typedef struct pid8_s {
 
-    uint16_t yaw_lpf_hz;                    // Additional yaw filter when yaw axis too noisy
-    uint16_t dterm_lpf_hz;                  // Delta Filter in hz
+    uint16_t yaw_lowpass_hz;                    // Additional yaw filter when yaw axis too noisy
```
Breaking PG here ...

ledvinap commented on src/main/interface/settings.c:
```diff
@@ -637,7 +637,8 @@ const clivalue_t valueTable[] = {
 // PG_PID_PROFILE
     { "dterm_lowpass_type",         VAR_UINT8  | PROFILE_VALUE | MODE_LOOKUP, .config.lookup = { TABLE_LOWPASS_TYPE }, PG_PID_PROFILE, offsetof(pidProfile_t, dterm_filter_type) },
-    { "dterm_lowpass",              VAR_INT16  | PROFILE_VALUE, .config.minmax = { 0, 16000 }, PG_PID_PROFILE, offsetof(pidProfile_t, dterm_lpf_hz) },
+    { "dterm_lowpass_hz",           VAR_INT16  | PROFILE_VALUE, .config.minmax = { 0, 16000 }, PG_PID_PROFILE, offsetof(pidProfile_t, dterm_lowpass_hz) },
```
Dump/restore will fail ...

DieHertz replied:
That's not an issue, we rename some parameters in every version, to finally achieve uniformity.

flint723 commented:
@ledvinap  I get what you are saying about the passband getting halved which hurts flight performance.  I noticed it right away when flying those logs shown above but did not realized what was happening.  I kept raising the higher cutoff and it wasn't changing how it flew at all.

I've now flown with Dterm pt1 cutoff at 120hz, 500hz which is much better than 100, 500hz.  And now 130hz, 500hz which I have to say it flies smoother, less prop wash, and far less dterm noise above 200hz when compared to my best flying "conventional" setting (which is flight 85 in the above pictures logs).

ledvinap commented:
@flint723 : Did you try Butterworth around 200Hz ?

ledvinap commented:
Also ... passband isn't halved, filter quality factor is 1/2. 
With 2nd order filter, you can get some boost (from resonance) at cutoff frequency, making passband flatter. This is impossible when simply chaining PT1s

flint723 commented:
Sorry I don't know filter terminology... or how to compile my own hex so no I haven't tried Butterworth.

flint723 commented:
I definitely would try it vs dual pt1 if I could get a OBF4 hex.

flint723 commented:
After more flights testing different combinations, I have some that fly equivalent or slightly better in terms of prop wash and overall handling (hard to say without flying one for 5+ packs then the other) to conventional filtering, but have far less Dterm noise above 200hz.

Pictured are some of the better combos I found for my setup, after all the other flights I re-flew the best ones followed by my conventional setup in flight 106 (gyro soft lowpass = 100hz, dynamic on, Dterm PT1 100hz, no dterm notch).  Dynamic notch is on in all of these, stage2 BQRC is off.  MPU6000 @8k8k Dshot1200.

What I like about the dual stage, is it allows you to tune higher or lower frequency exactly how you want.  My current favorite was flight 105 (gyro soft lowpass = 100hz, Dterm PT1 130hz and 500hz).  After trying many combinations I kept coming back to that one.  Right behind it was flight 104 (gyro soft lowpass = 90hz, Dterm PT1 140hz and 500hz).  Flight 103 flew quite well, but as you can see is filtered more and showed just a little bit more prop wash but much less motor noise.  If I was blind testing flight 105 and 106 I don't think I could tell the difference, except possibly by sound.

In theory both flight 105 and 104 have very similar latency on dterm (at 50hz) to a single PT1 set to 100hz cutoff.

![flight 103-106 dterm](https://user-images.githubusercontent.com/30608240/37237849-941ab5d6-23df-11e8-87b9-f51ce702db3b.jpg)
![flight 103-106 motors](https://user-images.githubusercontent.com/30608240/37237860-9f776938-23df-11e8-8d18-2eb97974d3df.jpg)

List of the settings I tried:
![image](https://user-images.githubusercontent.com/30608240/37237877-ecb2dfac-23df-11e8-9738-1ebcdf40dced.png)
And the logs themselves:
https://www.dropbox.com/s/usimc5c9m5mnkn8/Dterm_lowpass2%20logs%203-9-2018.zip?dl=0

ctzsnooze commented:
I now have very positive feedback from five extremely good pilots, all using this build.

Consensus is that noise is much better attenuated, especially on D, allowing higher set points and better performance with cooler motors.  As Flint says, it's now very easy to fine tune P or D filtering to suit the intent of the quad.

@fujin @DieHertz - are there any outstanding issues here that would stop this going into betaflight 3.4?

I am confident that once you fly this build you'll want the D defaults changed to those I recommended previously.

PS In relation to which spectrums to use when evaluating noise, I find motors to be least useful.  That's because they integrate noise from both axes, and the axes can peak differently, and the motor spectrum doesn't tell you whether P or D is source of the noise, or which filter to adjust.  

So I pick an axis, doesn't matter which, and compare the P noise spectrum (affected only by gyro filters) vs the D noise spectrum (affected by gyro filters first then more specifically by d filters).  D typically gets less weighting into PIDsum, so its good to compare the P and D noise spectrums to pidSum spectrum, so as to see which contributes most.  

Then, whichever of P or D contributes the majority of the noise into PIDsum will determine where to put your filtering efforts.  Typically D contributes the most to pidSum noise, by far.  Hence, if noise reduction is the goal, a finding like that suggests that it would be best to filter D harder in that case.

DieHertz commented:
Looks fine by me, subsequent refactorings could be done by us in further PRs.

ctzsnooze commented:
Thanks, @DieHertz.  I do like the consistent parameter naming, the CLI is now much better.  
What about blackbox, OSD and other complexity of that kind?

ctzsnooze commented:
PS, if you haven't flown it, these settings are good conservative starting points:

`set gyro_lowpass_type = PT1
set gyro_stage2_lowpass_hz = 250
set gyro_lowpass_hz = 120
set gyro_notch1_hz = 0
set gyro_notch1_cutoff = 0
set gyro_notch2_hz = 0
set gyro_notch2_cutoff = 0

set dterm_lowpass_type = PT1
set dterm_lowpass_hz = 80
set dterm_lowpass2_hz = 160
set dterm_notch_hz = 0
set dterm_notch_cutoff = 0
set yaw_lowpass_hz = 0

feature DYNAMIC_FILTER`

ctzsnooze commented:
@mikeller what do I need to do next?

mikeller commented:
@ctzsnooze: Thanks for this. Can you please rebase this to get rid of the conflict?

@fujin: I assume that this will not interfere with #5391, since the changes are to PID filters?

fujin commented:
@mikeller yes, should not conflict with #5391

ctzsnooze commented:
Sorry I don't know how to rebase.

DieHertz commented:
I could help and guide you if you contact me over Slack

DieHertz commented:
You did a merge instead of rebase.

ctzsnooze commented:
I'm joining slack now - I did try to rebase, sorry.  Maybe should do it all again from scratch.

DieHertz commented:
Why have you closed it? There was no need to

ctzsnooze commented:
Sorry I closed it because I stuffed up the merge... I have now fixed that with a new pull here:

https://github.com/betaflight/betaflight/pull/5458

I couldn't get the slack email reply and its 1am now.  Sorry of making a mess.  But the new pull should be OK I fixed the one line that needed changing.

Apologies.

atomiclama commented:
@ctzsnooze Just like to say thanks for this. On my ageing CC3D the motors are now cooler and the quad sounds smoother. Can't do logs to confirm but feels better.
